clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2070,4 +2086,274 @@ void handleFailure(Throwable throwable) {
+                    for (Map.Entry<String, KafkaFutureImpl<ConsumerGroupDescription>> entry : consumerGroupFutures.entrySet()) {
+                        final String groupId = entry.getKey();
+                        final KafkaFutureImpl<ConsumerGroupDescription> future = entry.getValue();
+                        final DescribeGroupsResponse.GroupMetadata groupMetadata = response.groups().get(groupId);
+                        final Errors groupError = groupMetadata.error();
+                        if (groupError != Errors.NONE) {
+                            future.completeExceptionally(groupError.exception());
+                            continue;
+                        }
+
+                        final String protocolType = groupMetadata.protocolType();
+                        if (protocolType.equals(ConsumerProtocol.PROTOCOL_TYPE) || protocolType.isEmpty()) {
+                            final List<DescribeGroupsResponse.GroupMember> members = groupMetadata.members();
+                            final List<MemberDescription> consumers = new ArrayList<>(members.size());
+
+                            for (DescribeGroupsResponse.GroupMember groupMember : members) {
+                                final PartitionAssignor.Assignment assignment =
+                                    ConsumerProtocol.deserializeAssignment(
+                                        ByteBuffer.wrap(Utils.readBytes(groupMember.memberAssignment())));
+
+                                final MemberDescription memberDescription =
+                                    new MemberDescription(
+                                        groupMember.memberId(),
+                                        groupMember.clientId(),
+                                        groupMember.clientHost(),
+                                        new MemberAssignment(assignment.partitions()));
+                                consumers.add(memberDescription);
+                            }
+                            final String protocol = groupMetadata.protocol();
+                            final ConsumerGroupDescription consumerGroupDescription =
+                                new ConsumerGroupDescription(groupId, protocolType.isEmpty(), consumers, protocol);
+                            future.complete(consumerGroupDescription);
+                        }
+                    }
+                }
+
+                @Override
+                void handleFailure(Throwable throwable) {
+                    completeAllExceptionally(consumerGroupFutures.values(), throwable);
+                }
+            }, nowDescribeConsumerGroups);
+
+        }
+
+        return new DescribeConsumerGroupsResult(new HashMap<String, KafkaFuture<ConsumerGroupDescription>>(consumerGroupFutures));
+    }
+
+    @Override
+    public ListConsumerGroupsResult listConsumerGroups(ListConsumerGroupsOptions options) {
+        final Collection<Node> nodes = listNodes(options.timeoutMs());
```

Comment:
This is not correct.  It's blocking, which turns this into a block API rather than a non-blocking one.

Reply:
+1